components/brave_rewards/browser/diagnostic_log.h

```diff
@@ -0,0 +1,64 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_REWARDS_BROWSER_DIAGNOSTIC_LOG_H_
+#define BRAVE_COMPONENTS_BRAVE_REWARDS_BROWSER_DIAGNOSTIC_LOG_H_
+
+#include <string>
+
+#include "base/files/file.h"
+#include "base/files/file_path.h"
+#include "base/sequence_checker.h"
+#include "base/sequenced_task_runner.h"
+
+namespace brave_rewards {
+
+// This class provides access to a diagnostic log file. If the file
+// size ever exceeds the provided maximum file size, it is trimmed to
+// keep only the last n lines.
+class DiagnosticLog : public base::SupportsWeakPtr<DiagnosticLog> {
+ public:
+  DiagnosticLog(const base::FilePath& path,
+                const int64_t max_file_size,
+                const int keep_num_lines);
+  DiagnosticLog(const DiagnosticLog&) = delete;
+  DiagnosticLog& operator=(const DiagnosticLog&) = delete;
+  ~DiagnosticLog();
+
+  using ReadCallback = base::OnceCallback<void(const std::string& data)>;
+  using WriteCallback = base::OnceCallback<void(const bool result)>;
+
+  // Reads last |num_lines| lines of file. If |num_lines| is -1, reads
+  // the entire file.
+  void ReadLastNLines(const int num_lines, ReadCallback callback);
+
+  // Appends |log_entry| to end of file. If file doesn't exist, it is
+  // created. If total file size exceeds |max_file_size|, removes all
+  // but the last |keep_num_lines| lines.
+  void Write(const std::string& log_entry, WriteCallback callback);
+
+  // Formats |message| in a user-friendly manner.
+  static std::string FriendlyFormatLogEntry(const base::Time& time,
```

Comment:
Since callers always call this before calling `Write`, it would also work to create a `Write` overload with this signature and do the formatting internally. Up to you though!

Reply:
Ah yeah, that would be much nicer, thanks.